datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,969 @@
+        (!self.desc && new_val > *worst_val) || (self.desc && new_val < *worst_val)
+    }
+
+    fn worst_map_idx(&self) -> usize {
+        self.heap.worst_map_idx()
+    }
+
+    fn renumber(&mut self, heap_to_map: &[(usize, usize)]) {
+        self.heap.renumber(heap_to_map);
+    }
+
+    fn insert(&mut self, row_idx: usize, map_idx: usize, map: &mut Vec<(usize, usize)>) {
+        let vals = self.owned.as_primitive::<VAL>();
+        let new_val = vals.value(row_idx);
+
+        if self.heap.is_full() {
+            self.heap.replace_root(new_val, map_idx, map);
+        } else {
+            self.heap.append(new_val, map_idx, map);
+        }
+    }
+
+    fn replace_if_best(
+        &mut self,
+        heap_idx: usize,
+        row_idx: usize,
+        map: &mut Vec<(usize, usize)>,
+    ) {
+        let vals = self.owned.as_primitive::<VAL>();
+        let new_val = vals.value(row_idx);
+        self.heap.replace_if_best(heap_idx, new_val, map);
+    }
+
+    fn take_all(&mut self, heap_idxs: Vec<usize>) -> ArrayRef {
+        let vals = self.heap.take_all(heap_idxs);
+        let vals = Arc::new(PrimitiveArray::<VAL>::from_iter_values(vals));
+        adjust_output_array(&self.data_type, vals).expect("Type is incorrect")
+    }
+}
+
+pub struct StringHashTable {
+    owned: ArrayRef,
+    map: CustomMap<Option<String>>,
+    rnd: RandomState,
+}
+
+impl StringHashTable {
+    pub fn new(limit: usize) -> Self {
+        let vals: Vec<&str> = Vec::new();
+        let owned = Arc::new(StringArray::from(vals));
+        Self {
+            owned,
+            map: CustomMap::new(limit, limit * 10),
+            rnd: ahash::RandomState::default(),
+        }
+    }
+}
+
+impl LimitedHashTable for StringHashTable {
+    fn set_batch(&mut self, ids: ArrayRef) {
+        self.owned = ids;
+    }
+
+    fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    unsafe fn update_heap_idx(&mut self, mapper: &[(usize, usize)]) {
+        self.map.update_heap_idx(mapper);
+    }
+
+    unsafe fn heap_idx_at(&self, map_idx: usize) -> usize {
+        self.map.heap_idx_at(map_idx)
+    }
+
+    fn drain(&mut self) -> (ArrayRef, Vec<usize>) {
+        let (ids, heap_idxs) = self.map.drain();
+        let ids = Arc::new(StringArray::from(ids));
+        (ids, heap_idxs)
+    }
+
+    unsafe fn find_or_insert(
+        &mut self,
+        row_idx: usize,
+        replace_idx: usize,
+        mapper: &mut Vec<(usize, usize)>,
+    ) -> (usize, bool) {
+        let ids = self
+            .owned
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .expect("StringArray required");
+        let id = if ids.is_null(row_idx) {
+            None
+        } else {
+            Some(ids.value(row_idx))
+        };
+
+        let hash = self.rnd.hash_one(id);
+        if let Some(map_idx) = self
+            .map
+            .find(hash, |mi| id == mi.as_ref().map(|id| id.as_str()))
```

Comment:
`&str` comparison means we don't `.to_string()` until we have to. Putting this in a type-specific `impl` means we can do the same for things like `Row`, not invoking `.to_owned()` until we have to.